graph.py

```diff
@@ -0,0 +1,125 @@
+        return result
+
+    def add_node(self, n):
+        """
+        Add a node to the Graph
+        """
+        self.g[n] = []
+
+    def add_edge(self, n1, n2):
+        """
+        Add an edge between two nodes
+
+        If any of the nodes were not created then they will be created
+        """
+        if not self.g.has_key(n1):
+            self.add_node(n1)
+        if not self.g.has_key(n2):
+            self.add_node(n2)
+        self.g[n1].append(n2)
+
+    def del_node(self, n):
+        """
+        Delete an existing node from the Graph
+        """
+        try:
+            del self.g[n]
+            for node in self.g.iterkeys():
+                if n in self.g[node]:
+                    self.g[node].remove(n)
+        except KeyError:
+            raise IndexError("Node doesn't exist")
+
+    def del_edge(self, n1, n2):
+        """
+        Delete an edge from the Graph. If an edge doesn't exist pass
+        """
+        try:
+            self.g[n1].remove(n2)
+        except ValueError:
+            pass
+        except KeyError:
+            raise IndexError("First node doesn't exist")
+
+    def has_node(self, n):
+        """
+        Return if Graph has a node
+        """
+        return self.g.has_key(n)
+
+    def neighbors(self, n):
+        """
+        Return the edges for a node in a list
+        """
+        try:
+            return self.g[n]
+        except KeyError:
+            raise IndexError("Node doesn't exist")
+
+    def adjacent(self, n1, n2):
+        """
+        Return if an edge exist between two nodes
+        """
+        try:
+            return n2 in self.g[n1]
+        except KeyError:
+            raise IndexError("One or more nodes doesn't exist")
+
+    def depth_first_traversal(self, node,result=""):
+        if not str(node) in result:
+            result = '{}{}'.format(result, node)
+            if self.g[node]:
+                for edge_node in self.g[node]:
+                    result = self.depth_first_traversal(edge_node,result)
+        return result
+
+    def breadth_first_traversal(self, node):
+        """
+        Iterate over a graph and return a path as a string.
+        """
+        children = self.g[node]
+
+        grandchildren, result = self.breadth_iterator(children, [], str(node))
+
+        while grandchildren:
+            children = grandchildren
+
+            grandchildren, result = self.breadth_iterator(children, [], result)
+
+        return result
+
+    def breadth_iterator(self, children, grandchildren, result):
+        """
+        Traverse a graph by generations.
+
+        Simultaneously grow a result string from list of current generations
+        and create a list of the next generation.
+        """
+        for child in children:
+            if str(child) not in result:
+                result = '{}{}'.format(result, child)
+                grandchildren += self.g[child]
+
+        return grandchildren, result
```

Comment:
Interesting approach to the problem here.  It works, but is not the most efficient method of solving the problem.  Here's the algorithm you have:

1.  get a list of the children of the current node.  
2.  get a list of the children of each of them and append it to a list called 'grandchildren'
3.  when you run out of children, make the list of grandchildren the new list of children and repeat.

This works, but at the cost of building a potentially large number of lists.  Especially since you're using concatenation (the `+` operator) to join your partial lists.  

Can you think of a data structure we've studied where you add things to a list and can process them out of that list in the same order they were added?  Could you see how to use such a structure to simplify your algorithm here?